File: vol1/ch1/LeetCode189.py
class Solution:
    def rotate(self, nums, k: int) -> None:
        n=len(nums)
        if n==1: return
        k=k%n
        self.reverse(nums,0,n-k-1)
        self.reverse(nums,n-k,n-1)
        self.reverse(nums,0,n-1)

    def reverse(self,nums,s,t):
        i,j=s,t
        while i<j:
            nums[i],nums[j]=nums[j],nums[i]
            i+=1; j-=1



class Solution:
    def rotate(self, nums,k:int):
        n=len(nums)
        if n==1: return
        k=k%n
        tmp=nums[n-k-1::-1]+nums[n-1:n-k-1:-1]
        tmp.reverse()
        nums[:]=tmp[:]

File: vol1/ch1/LeetCode189.cpp
class Solution {
public:
    void rotate(vector<int>& nums,int k) {
        int n=nums.size();
        if(n==1) return;
        k=k%n;
        reverse(nums,0,n-k-1);
        reverse(nums,n-k,n-1);
        reverse(nums,0,n-1);
    }

    void reverse(vector<int>&nums,int s,int t) {
        int i=s,j=t;
        while(i<j) {
            swap(nums[i],nums[j]);
            i++; j--;
        }
    }
};

